Bitbucket.Authentication/Authentication.cs

```diff
@@ -137,21 +152,38 @@ public Credential GetCredentials(TargetUri targetUri, string username)
         /// <param name="targetUri"></param>
         /// <param name="username"></param>
         /// <returns></returns>
-        private TargetUri GetPerUserTargetUri(TargetUri targetUri, string username)
+        public TargetUri GetPerUserTargetUri(TargetUri targetUri, string username)
         {
-            if (string.IsNullOrWhiteSpace(username))
+            // belt and braces, don't add a username if the URI already contains one.
+            if (string.IsNullOrWhiteSpace(username) || TargetUriContainsUsername(targetUri))
             {
                 return targetUri;
             }
 
             return new TargetUri(targetUri.ActualUri.AbsoluteUri.Replace(targetUri.Host, username + "@" + targetUri.Host));
         }
 
+        public TargetUri GetHostTargetUri(TargetUri targetUri)
```

Comment:
Are we better of adding this functionality directly to the `TargetUri` type instead of containing it in the `BitBucket.Authentication` type? It seems to me that it could benefit the other authentication types to have access to these routines.